imdsclient: add retry to fetch_token

jpculp opened the pull request:
**Description of changes:**

This adds retry logic if `imdsclient` fails to fetch a session token.

Also includes small changes to the `fetch_imds` retry logic, such as moving
the sleep after the ensure `imds_attempts <= max_imds_attempts` and bumping
the wait time from 100ms to 1s.

**Testing done:**

- Built `aws-k8s-1.20` ami and launched instance.
- Instance connected to eks cluster.
- Connected to control container via ssm session.
- Verified that `host-containers.admin.user-data` contained a base64-encoded block.
- Connected to admin container via ssh.
- Verified that `/.bottlerocket/host-containers/admin/user-data` contained JSON.
- Ran `sudo sheltie` to verify root shell was still available.
- Checked for failed systemd units.
- Ran `pluto` with it's sub-commands to verify functionality.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
imdsclient: add retry to fetch_token

This adds retry logic if imdsclient fails to fetch a session token.

Also includes small changes to the fetch_imds retry logic, such as
moving the sleep after the ensure imds_attempts <= max_imds_attempts and
bumping the wait time from 100ms to 1s.

bcressey commented on sources/imdsclient/src/lib.rs:
```diff
@@ -351,23 +354,39 @@ fn build_public_key_targets(public_key_list: &str) -> Vec<String> {
-        uri,
-        code,
-    })
+    let mut token_attempts: u8 = 0;
+    let max_token_attempts: u8 = 2;
+    loop {
+        token_attempts += 1;
+        ensure!(
+            token_attempts <= max_token_attempts,
+            error::FailedFetchToken { token_attempts }
+        );
+        if token_attempts > 1 {
+            time::sleep(Duration::from_secs(5)).await;
+        }
```
I'd like to align the number of retries and the length of the delay with `fetch_imds`  - the requests are going to the same service and reaching the failure state in either function will likely result in a boot failure.

jpculp replied:
The retries in `fetch_imds` are mostly to trigger token refreshes in the even the token is expired as opposed to IMDS unavailability (hence a shorter sleep). Each client initialization and `fetch_imds` attempt is going to fetch a new token. If IMDS is unreachable for whatever reason, the failure would occur during the `fetch_token`. Since we are adding retries to the token fetch now it might make sense to drop the max_imds_attempts from 3 to 2, but I wouldn't recommend aligning the sleep duration.

jpculp replied:
Instead of dropping `fetch_imds` from 3 to 2, I bumped `fetch_token` from 2 to 3.

zmrow commented:
Once @bcressey 's concern is handled, lgtm

jpculp commented:
- Replaced variables like `imds_attempts` with `attempt`.
- Increased the `fetch_token` max attempts from 2 to 3. Now both `fetch_token` and `fetch_imds` have 3 attempts.
- Changed wording for `fetch_token` error message to be more descriptive.
